### PR TITLE
add support for multipart/form-data POST,PUT,PATCH similar to httpbin's

### PR DESCRIPTION
**Pull request by jordan0day**
i think the output matches httpbin's behavior pretty closely for multipart/form-data requests. supports both regular form-data and file uploads.

### CODE REVIEW

File: lib/httparrot/p_handler.ex [hunk 1]
```diff
@@ -57,4 +58,70 @@ defmodule HTTParrot.PHandler do
...
+
+    post(req, [form: form_parts, files: file_parts, data: "", json: nil])
+  end
+
+  defp handle_multipart(req, parts \\ []) do
+    case :cowboy_req.part(req) do
+      {:done, req} -> {:ok, parts, req}
+      {:ok, headers, req} ->
+        content_disposition = List.keyfind(headers, "content-disposition", 0)
+        if content_disposition != nil do
```
Comment (edgurgel) [inline lib/httparrot/p_handler.ex]: I think you can use just `if content_disposition do` here

**Comment by edgurgel** [comment]
That's a great contribution! Thank you very much @jordan0day! 

The build is failing for other reasons, I will just drop support to older Elixir versions.

**Comment by edgurgel** [comment]
Also thanks for the extensive unit tests! :+1:

File: lib/httparrot/p_handler.ex [hunk 2]
```diff
@@ -57,4 +58,70 @@ defmodule HTTParrot.PHandler do
...
+
+  def post_multipart(req, _state) do
+    {:ok, parts, req} = handle_multipart(req)
+
+    filter = fn({type, _name, _body}, type_atom) -> type == type_atom end
+    reducer = fn({_type, name, body}, acc) -> acc ++ [{name, body}] end
+
+    # the other post handlers return a list with a single empty tuple if
+    # there's no data for forms, so let's match that behavior...
+    normalize = fn(parts) -> if parts == [], do: [{}], else: parts end
```
Comment (edgurgel) [inline lib/httparrot/p_handler.ex]: I think this can be a private function on the module.

File: lib/httparrot/p_handler.ex [hunk 3]
```diff
@@ -57,4 +58,70 @@ defmodule HTTParrot.PHandler do
...
+    end
+  end
+
+  defp parse_content_disposition_header(header) do
+    parts = elem(header, 1) |> String.split(";")
+
+    type = Enum.at(parts, 0)
+    parts = Enum.drop(parts, 1)
+
+    Enum.reduce(parts, %{:type => type}, fn part, acc ->
```
Comment (edgurgel) [inline lib/httparrot/p_handler.ex]: I think you can use for comprehension here to be more concise:

```elixir
    for part <- parts, into: %{type: type} do
      [key, value] = String.split(part, "=")
      key = String.strip(key)
      value = String.strip(value) |> String.replace("\"", "")
      {key, value}
    end
```

What do you think?

File: lib/httparrot/p_handler.ex (hunk 2 above)
Comment (jordan0day) [inline lib/httparrot/p_handler.ex]: Are you referring to post_multipart, or the filter,reducer, normalize anonymous fn's?

Comment (edgurgel) [inline lib/httparrot/p_handler.ex]: Just the normalize one.

**Comment by jordan0day** [comment]
Updated to use comprehensions in a couple of places -- both where you suggested in the parse_content_disposition_header fn and in the main post_multipart header, removing the filter and reducer anonymous fn's.

File: lib/httparrot/p_handler.ex [hunk 4]
```diff
@@ -57,4 +58,61 @@ defmodule HTTParrot.PHandler do
...
+
+    for part <- parts, into: %{} do
+      case String.split(part, "=") |> Enum.map(&String.strip/1) do
+        [type] -> {:type, type}
+        [key, value] -> {key, String.replace(value, "\"", "")}
+      end
+    end
+  end
+
+  defp normalize_list(list) when list == [], do: [{}]
```
Comment (edgurgel) [inline lib/httparrot/p_handler.ex]: I think you can use just ` defp normalize_list([])` without the `when` guard

**Comment by edgurgel** [comment]
Thanks for the update. I just commented a small thing. Again, thank you for the contribution :+1:

**Comment by jordan0day** [comment]
Ah yes, of course, good catch. Fixed.

**Pull request merged**